fix(router): hide emptyoutlet from consumers

daem0ndev opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
Currently, the activate events and component getter on the router outlet will return an `EmptyOutletComponent` when using lazy loaded components on named outlets.

Issue Number: https://github.com/angular/angular/issues/27696


## What is the new behavior?
The new behavior is that the activate events as well as component get correctly return the lazy loaded component instead of an `EmptyOutletComponent`

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

daem0ndev commented:
@jasonaden requesting a review at your convenience

jasonaden commented:
[Presubmit](http://test/OCL:244879039:BASE:244879069:1556040544277:6ce7ecd4)

jasonaden commented:
Google testing fails due to cyclical dependencies. This is the cycle of files referenced:

```
router/src/directives/router_outlet.ts 
router/public_api.ts 
router/src/index.ts 
router/src/directives/router_link.ts 
router/src/router.ts 
router/src/config.ts 
router/src/components/empty_outlet.ts 
router/src/directives/router_outlet.ts
```

I can help debug a bit more after RC, but if you can fix it sooner, I can re-run internal CI checks (the "Presubmit" mentioned above).

daem0ndev commented:
Thanks! I’ll take a look

jasonaden commented on packages/router/src/directives/router_outlet.ts:
```diff
@@ -7,12 +7,15 @@
  */
 
 import {Attribute, ChangeDetectorRef, ComponentFactoryResolver, ComponentRef, Directive, EventEmitter, Injector, OnDestroy, OnInit, Output, ViewContainerRef} from '@angular/core';
+import {takeUntil} from 'rxjs/operators';
```
Let's see if we can do this without adding `takeUntil` operator dependency.

daem0ndev replied:
Done: I used a simple subscription / unsubscribe on destroy instead of takeUntil

jasonaden commented on packages/router/src/components/empty_outlet.ts:
```diff
@@ -6,7 +6,13 @@
+import {Subject} from 'rxjs';
+
+import {Data} from '../config';
+import {RouterOutlet} from '../directives/router_outlet';
```
Due to circular dependency, let's refactor to move `EmptyOutletComponent` into `router_outlet.ts` file. Include a comment on why this is done.

daem0ndev replied:
Done.

jasonaden reviewed (changes requested):
Thanks for the PR! Let's get this merged soon.

daem0ndev committed:
refactor(router): move empty outlet component

Due to future circular references between RouterOutlet and EmptyOutletComponent, the EmptyOutletComponent has been moved to the RouterOutlet file.

daem0ndev committed:
fix(router): hide `EmptyOutletComponent` from consumers

This commit hides the internal-only `EmptyOutletComponent` from
consumers.

Currently, the activate events and component getter on the router outlet will
return an `EmptyOutletComponent` when using lazy loaded components on named outlets.
This was not intended to be the case. Instead, it was meant to be a
stand-in outlet just to make the routing work correctly. It should emit
the information from its child instead; that is the outlet that is
activated from the actual `Route` in the config.

Fixes #27696

BREAKING CHANGE: This breaking change applies only to named routes with
children but no component. These routes are automatically assigned a placeholder
component with a `RouterOutlet` for rendering the children routes.
This internal-only placeholder would
previously be exposed by the parent `RouterOutlet` properties
(`component` and the `activate` event) and subsequently hides the
information about the activated child routes.
The new behavior for this placeholder is to act as a passthrough for the
child activated routes. The parent `RouterOutlet` will now return the
information from the child `RouterOutlet` of the placeholder rather than
the information of the placeholder component. This change also affects
the timing of the `activate` event of the parent because it will emit once the
_child_ outlet activates rather than when the placeholder component is
created.

jasonaden commented:
[New presubmit](https://test.corp.google.com/ui#id=OCL:244879039:BASE:244879069:1556040544279:fcb6e7a3)

daem0ndev commented:
@jasonaden could I get some info on the google3 failure for `requires-piper-patch`, all tests seem to be passing for me locally so I am not sure what is failing on the g3 side

billdwhite commented:
Is there an update on this?

daem0ndev commented:
@jasonaden *bump* could I get some eyes on this?

Marka83 commented:
Can anyone take a look at this?
If @daem0ndev invested his time to fix this, at least tell him what's the problem with the tests.

billdwhite commented:
This issue is just killing me;  any way we can get this pull request reviewed???

lucaswxp commented:
Almost a year, any workaround this?

atscott commented on packages/router/src/directives/router_outlet.ts:
```diff
@@ -140,7 +147,18 @@ export class RouterOutlet implements OnDestroy, OnInit {
+    // from the inner outlet
+    const c = this.activated.instance;
+    if (c instanceof ɵEmptyOutletComponent) {
+      const sink = new Subscription;
```
```suggestion
      const sink = new Subscription();
```

nit: Never invoke a constructor in a `new` statement without using parentheses `()`.
https://google.github.io/styleguide/jsguide.html#disallowed-features-omitting-parents-with-new

atscott commented on packages/router/src/directives/router_outlet.ts:
```diff
@@ -121,7 +126,9 @@ export class RouterOutlet implements OnDestroy, OnInit {
       this.activated = null;
       this._activatedRoute = null;
-      this.deactivateEvents.emit(c);
+      if (!(c instanceof ɵEmptyOutletComponent)) {
```
How about adding a comment here saying that `EmptyOutletComponent` deactivateEvents are subscribed to and emitted in `activateWith` instead of here?

atscott reviewed (commented):
FYI - I rebased this PR so [presubmit](http://test/OCL:301607837:BASE:301609480:1584550656741:9d603cd1) could be run again.

atscott reviewed (commented):
Also looks like the circular deps is still an issue:
```
 • packages/router/src/config.ts → packages/router/src/directives/router_outlet.ts
 • packages/router/src/directives/router_outlet.ts → packages/router/src/router_outlet_context.ts
 • packages/router/src/config.ts → packages/router/src/directives/router_outlet.ts → packages/router/src/router_outlet_context.ts → packages/router/src/router_state.ts
 • packages/router/src/config.ts → packages/router/src/directives/router_outlet.ts → packages/router/src/router_outlet_context.ts → packages/router/src/router_state.ts → packages/router/src/shared.ts
  ```

You could move `standardizeConfig` from `config.ts` to a new `config_utils.ts` (and potentially the other functions in `config.ts`). This would also help to separate the public API types and interfaces from the internal utility functions and classes.

daem0ndev commented:
@atscott thank you soooo much for taking a look! I will work on these updates asap

atscott commented:
[global](https://test.corp.google.com/ui#id=OCL:301607837:BASE:301831157:1584635402186:f11b8230) presubmit ran and no failures as a result of this PR so it should be good to go once CI is green.

AndrewKushnir commented:
Hi @daem0ndev, thanks for creating this PR!

Just want to give a quick update. We've [landed a change](https://github.com/angular/angular/commit/4dfc3fe6a29df0183aea9b946379157f3418a6e5) similar to the one described by @atscott in https://github.com/angular/angular/pull/29746#pullrequestreview-377108598, so we also rebased this PR to leverage that change. I also added a [fixup commit](https://github.com/angular/angular/pull/29746/commits/e6aeaafde20369f556bbe52c4e550b68d01c3f4f) to your PR to address the feedback from @atscott.

After re-running CI with the most up-to-date version of this PR (based on the latest version of master branch), we noticed a size regression in one of the test app (see [this CI job output](https://app.circleci.com/pipelines/github/angular/angular/19291/workflows/ef3165e4-b1e1-405e-b708-3b4a45cbed02/jobs/775850)). It looks like referencing the `ɵEmptyOutletComponent` directly makes it non-tree-shakable. We'll look into possible solutions and let you know.

Thank you.

daem0ndev commented:
Awesome! Thanks so much for jumping in on this. If there’s anything else I can do from my end to help, let me know!

falconeta commented:
Hi @daem0ndev, any news about the fix?

AndrewKushnir commented:
There is no progress on this change yet. The problem is that the code introduces the use of the `ViewChild` decorator, which was not present before in the Router package. That makes the `ViewChild` symbol (along with some underlying code to support that) non-tree-shakable, which means that this extra code would always be included into prod bundles (which is undesirable). We'd need to perform further research to find a way to implement the same logic in a more tree-shakable fashion, but there is no ETA at this moment.

mary-poppins commented:
You can preview e702301 at https://pr29746-e702301.ngbuilds.io/.
You can preview 0035c9d at https://pr29746-0035c9d.ngbuilds.io/.
You can preview 86d2279 at https://pr29746-86d2279.ngbuilds.io/.
You can preview 5898c29 at https://pr29746-5898c29.ngbuilds.io/.
You can preview ea0535f at https://pr29746-ea0535f.ngbuilds.io/.
You can preview 9859d49 at https://pr29746-9859d49.ngbuilds.io/.
You can preview 091c96b at https://pr29746-091c96b.ngbuilds.io/.
You can preview 857df54 at https://pr29746-857df54.ngbuilds.io/.
You can preview 97ea5aa at https://pr29746-97ea5aa.ngbuilds.io/.
You can preview 85d0693 at https://pr29746-85d0693.ngbuilds.io/.
You can preview 06262a7 at https://pr29746-06262a7.ngbuilds.io/.
You can preview e6aeaaf at https://pr29746-e6aeaaf.ngbuilds.io/.

mary-poppins commented:
You can preview d9b82f4 at https://pr29746-d9b82f4.ngbuilds.io/.

mary-poppins commented:
You can preview 9e85d17 at https://pr29746-9e85d17.ngbuilds.io/.

mary-poppins commented:
You can preview ef7a577 at https://pr29746-ef7a577.ngbuilds.io/.

mary-poppins commented:
You can preview 4fd5466 at https://pr29746-4fd5466.ngbuilds.io/.

mary-poppins commented:
You can preview 9c9b719 at https://pr29746-9c9b719.ngbuilds.io/.

mary-poppins commented:
You can preview f54b5e6 at https://pr29746-f54b5e6.ngbuilds.io/.

mary-poppins commented:
You can preview 3dd2678 at https://pr29746-3dd2678.ngbuilds.io/.

atscott commented:
Changing the implementation here to avoid `@ViewChild` to get the `RouterOutlet` in `EmptyOutletComponent` but instead do `@Optional` to query for the parent `EmptyOutletComponent` in `RouterOutlet` resulted in a much smaller bundle size increase: 1500B vs the [6200B from before](https://app.circleci.com/pipelines/github/angular/angular/19291/workflows/ef3165e4-b1e1-405e-b708-3b4a45cbed02/jobs/775850)

@AndrewKushnir  I think this should be ready for review again.

mary-poppins commented:
You can preview d9406eb at https://pr29746-d9406eb.ngbuilds.io/.

mary-poppins commented:
You can preview 72c6fd4 at https://pr29746-72c6fd4.ngbuilds.io/.

mary-poppins commented:
You can preview 5c8725b at https://pr29746-5c8725b.ngbuilds.io/.

mary-poppins commented:
You can preview f541c5d at https://pr29746-f541c5d.ngbuilds.io/.

mary-poppins commented:
You can preview 95df0de at https://pr29746-95df0de.ngbuilds.io/.

mary-poppins commented:
You can preview 1ccec5d at https://pr29746-1ccec5d.ngbuilds.io/.

mary-poppins commented:
You can preview c74bd8f at https://pr29746-c74bd8f.ngbuilds.io/.

mary-poppins commented:
You can preview 9ffef0a at https://pr29746-9ffef0a.ngbuilds.io/.

mary-poppins commented:
You can preview a30e391 at https://pr29746-a30e391.ngbuilds.io/.

mary-poppins commented:
You can preview d1b4a8b at https://pr29746-d1b4a8b.ngbuilds.io/.

mary-poppins commented:
You can preview 8d4fa14 at https://pr29746-8d4fa14.ngbuilds.io/.

AndrewKushnir commented on packages/router/src/directives/router_outlet.ts:
```diff
@@ -176,7 +177,8 @@ export class RouterOutlet implements OnDestroy, OnInit, RouterOutletContract {
       @Attribute('name') name: string, private changeDetector: ChangeDetectorRef,
-      private environmentInjector: EnvironmentInjector) {
+      private environmentInjector: EnvironmentInjector,
+      @Optional() parentEmptyOutlet?: ɵEmptyOutletComponent|null) {
```
It looks like we don't use it anywhere in the code...

atscott replied:
This must have been lost in the rebase... Have to figure out what it was used for...

atscott replied:
fixed

AndrewKushnir reviewed (approved):
LGTM, just one minor comment on potentially unused dependency.

mary-poppins commented:
You can preview 3a901b3 at https://pr29746-3a901b3.ngbuilds.io/.

AndrewKushnir commented on packages/router/src/directives/router_outlet.ts:
```diff
@@ -177,7 +177,11 @@ export class RouterOutlet implements OnDestroy, OnInit, RouterOutletContract {
+      private environmentInjector: EnvironmentInjector,
+      @Optional() parentEmptyOutlet?: ɵEmptyOutletComponent|null) {
+    if (parentEmptyOutlet) {
+      parentEmptyOutlet.outlet = this;
```
It looks like this code was added in a recent fixup commit. I think it'd be great to add a quick comment here to explain why we reset the `outlet` value.

atscott replied:
It's not resetting it but rather setting it for the first time. We're doing this because we wanted to avoid `@ViewChild`. Added a comment.

AndrewKushnir replied:
Thanks Andrew! This is super helpful 👍

mary-poppins commented:
You can preview a10ee25 at https://pr29746-a10ee25.ngbuilds.io/.

daem0ndev commented:
@atscott @AndrewKushnir thank you guys so much for putting in time on this issue! 🎉

dylhunn commented:
@AndrewKushnir @daem0ndev @atscott This seems to have been approved fairly recently. Is this something we intend to merge soon? If so, I can try to help get it in this week while I'm caretaking

dylhunn commented:
Also @daem0ndev would you mind doing one more rebase, for CI and eventual merging?

atscott commented:
@dylhunn There is a test failure in an AngularJS/Angular hybrid app that we have not been able to fix.

pkozlowski-opensource commented:
It seems like this PR got stalled for over a year, doesn't merge cleanly any more and it is not fit for merge any more. 

I'm going to close this PR for now as i'm trying to cleanup PR reviews queue a bit. I'm not sure how much framework changed and if the underlying issue is still relevant but if so it would require new PR.

atscott commented:
Reopening to try another approach to avoid the ViewChild query: using `createComponent`. This requires `RouterOutlet` to become a `Component` instead of a `Directive` so we'll have to see if that breaks anything.

atscott commented:
TGP identified one failure that was legitimate. The update here changes the timing of the activate event where it doesn't emit until the child outlet activates. This won't happen until the next change detection run. If there are subscriptions to `activate` that mutate state read in a template before the outlet component is defined in the template, it results in the `ExpressionChanged...` error. 

It's a pretty specific set of circumstances that results in failure, but we may still want to consider this a breaking change and wait until a major version to include.
